File: schemas/hr_agricole.py
"""
Schémas Pydantic pour la validation des données RH agricoles
"""

from pydantic import BaseModel, UUID4, Field, validator
from typing import Optional, List, Dict, Any
from datetime import date, datetime
from enum import Enum

class TypePersonnel(str, Enum):
    """Types de personnel agricole"""
    PERMANENT = "PERMANENT"
    SAISONNIER = "SAISONNIER"
    TEMPORAIRE = "TEMPORAIRE"
    STAGIAIRE = "STAGIAIRE"

class SpecialiteAgricole(str, Enum):
    """Spécialités agricoles"""
    CULTURE = "CULTURE"
    ELEVAGE = "ELEVAGE"
    MARAICHAGE = "MARAICHAGE"
    ARBORICULTURE = "ARBORICULTURE"
    MAINTENANCE = "MAINTENANCE"
    LOGISTIQUE = "LOGISTIQUE"

class NiveauCompetence(str, Enum):
    """Niveaux de compétence"""
    DEBUTANT = "DEBUTANT"
    INTERMEDIAIRE = "INTERMEDIAIRE"
    AVANCE = "AVANCE"
    EXPERT = "EXPERT"

class TypeCertification(str, Enum):
    """Types de certifications agricoles"""
    PHYTOSANITAIRE = "PHYTOSANITAIRE"
    SECURITE = "SECURITE"
    CONDUITE_ENGINS = "CONDUITE_ENGINS"
    BIO = "BIO"
    QUALITE = "QUALITE"

class CompetenceAgricoleBase(BaseModel):
    """Schéma de base pour une compétence agricole"""
    specialite: SpecialiteAgricole
    niveau: NiveauCompetence
    cultures: List[str] = Field(default_factory=list)
    equipements: List[str] = Field(default_factory=list)
    date_acquisition: date
    date_mise_a_jour: Optional[date]
    validite: Optional[date]
    commentaire: Optional[str]
    donnees_supplementaires: Dict[str, Any] = Field(default_factory=dict)

    @validator('validite')
    def validite_future(cls, v):
        if v and v < date.today():
            raise ValueError("La date de validité doit être dans le futur")
        return v

class CompetenceAgricoleCreate(CompetenceAgricoleBase):
    """Schéma pour la création d'une compétence agricole"""
    employe_id: UUID4

class CompetenceAgricoleUpdate(CompetenceAgricoleBase):
    """Schéma pour la mise à jour d'une compétence agricole"""
    pass

class CompetenceAgricoleInDB(CompetenceAgricoleBase):
    """Schéma pour une compétence agricole en DB"""
    id: UUID4
    employe_id: UUID4

    class Config:
        orm_mode = True

class CertificationAgricoleBase(BaseModel):
    """Schéma de base pour une certification agricole"""
    type_certification: TypeCertification
    organisme: str
    numero: Optional[str]
    date_obtention: date
    date_expiration: Optional[date]
    niveau: Optional[str]
    commentaire: Optional[str]
    donnees_supplementaires: Dict[str, Any] = Field(default_factory=dict)

    @validator('date_expiration')
    def expiration_future(cls, v):
        if v and v < date.today():
            raise ValueError("La date d'expiration doit être dans le futur")
        return v

class CertificationAgricoleCreate(CertificationAgricoleBase):
    """Schéma pour la création d'une certification agricole"""
    competence_id: UUID4
    document_id: Optional[UUID4]

class CertificationAgricoleUpdate(CertificationAgricoleBase):
    """Schéma pour la mise à jour d'une certification agricole"""
    document_id: Optional[UUID4]

class CertificationAgricoleInDB(CertificationAgricoleBase):
    """Schéma pour une certification agricole en DB"""
    id: UUID4
    competence_id: UUID4
    document_id: Optional[UUID4]

    class Config:
        orm_mode = True

class AffectationParcelleBase(BaseModel):
    """Schéma de base pour une affectation à une parcelle"""
    date_debut: date
    date_fin: Optional[date]
    role: str
    responsabilites: List[str] = Field(default_factory=list)
    objectifs: Dict[str, Any] = Field(default_factory=dict)
    restrictions_meteo: List[str] = Field(default_factory=list)
    equipements_requis: List[str] = Field(default_factory=list)
    commentaire: Optional[str]
    donnees_supplementaires: Dict[str, Any] = Field(default_factory=dict)

    @validator('date_fin')
    def date_fin_valide(cls, v, values):
        if v and 'date_debut' in values and v < values['date_debut']:
            raise ValueError("La date de fin doit être postérieure à la date de début")
        return v

class AffectationParcelleCreate(AffectationParcelleBase):
    """Schéma pour la création d'une affectation"""
    employe_id: UUID4
    parcelle_id: UUID4

class AffectationParcelleUpdate(AffectationParcelleBase):
    """Schéma pour la mise à jour d'une affectation"""
    pass

class AffectationParcelleInDB(AffectationParcelleBase):
    """Schéma pour une affectation en DB"""
    id: UUID4
    employe_id: UUID4
    parcelle_id: UUID4

    class Config:
        orm_mode = True

class ConditionTravailAgricoleBase(BaseModel):
    """Schéma de base pour les conditions de travail agricole"""
    date: date
    temperature: Optional[float] = Field(None, ge=-50, le=50)
    humidite: Optional[int] = Field(None, ge=0, le=100)
    precipitation: bool = False
    vent: Optional[float] = Field(None, ge=0)
    exposition_soleil: Optional[int] = Field(None, ge=0)
    charge_physique: Optional[int] = Field(None, ge=1, le=10)
    equipements_protection: List[str] = Field(default_factory=list)
    incidents: List[Dict[str, Any]] = Field(default_factory=list)
    commentaire: Optional[str]
    donnees_supplementaires: Dict[str, Any] = Field(default_factory=dict)

class ConditionTravailAgricoleCreate(ConditionTravailAgricoleBase):
    """Schéma pour la création de conditions de travail"""
    employe_id: UUID4

class ConditionTravailAgricoleUpdate(ConditionTravailAgricoleBase):
    """Schéma pour la mise à jour de conditions de travail"""
    pass

class ConditionTravailAgricoleInDB(ConditionTravailAgricoleBase):
    """Schéma pour des conditions de travail en DB"""
    id: UUID4
    employe_id: UUID4

    class Config:
        orm_mode = True

class FormationAgricoleBase(BaseModel):
    """Schéma de base pour une formation agricole"""
    specialite: SpecialiteAgricole
    cultures_concernees: List[str] = Field(default_factory=list)
    equipements_concernes: List[str] = Field(default_factory=list)
    conditions_meteo: Dict[str, Any] = Field(default_factory=dict)
    pratiques_specifiques: List[str] = Field(default_factory=list)
    evaluation_terrain: bool = False
    resultats_evaluation: Dict[str, Any] = Field(default_factory=dict)
    commentaire: Optional[str]
    donnees_supplementaires: Dict[str, Any] = Field(default_factory=dict)

class FormationAgricoleCreate(FormationAgricoleBase):
    """Schéma pour la création d'une formation agricole"""
    formation_id: UUID4

class FormationAgricoleUpdate(FormationAgricoleBase):
    """Schéma pour la mise à jour d'une formation agricole"""
    pass

class FormationAgricoleInDB(FormationAgricoleBase):
    """Schéma pour une formation agricole en DB"""
    id: UUID4
    formation_id: UUID4

    class Config:
        orm_mode = True

class EvaluationAgricoleBase(BaseModel):
    """Schéma de base pour une évaluation agricole"""
    performances_cultures: Dict[str, Any] = Field(default_factory=dict)
    maitrise_equipements: Dict[str, Any] = Field(default_factory=dict)
    respect_securite: Dict[str, Any] = Field(default_factory=dict)
    adaptabilite_meteo: Dict[str, Any] = Field(default_factory=dict)
    gestion_ressources: Dict[str, Any] = Field(default_factory=dict)
    qualite_travail: Dict[str, Any] = Field(default_factory=dict)
    commentaire: Optional[str]
    donnees_supplementaires: Dict[str, Any] = Field(default_factory=dict)

class EvaluationAgricoleCreate(EvaluationAgricoleBase):
    """Schéma pour la création d'une évaluation agricole"""
    evaluation_id: UUID4

class EvaluationAgricoleUpdate(EvaluationAgricoleBase):
    """Schéma pour la mise à jour d'une évaluation agricole"""
    pass

class EvaluationAgricoleInDB(EvaluationAgricoleBase):
    """Schéma pour une évaluation agricole en DB"""
    id: UUID4
    evaluation_id: UUID4

    class Config:
        orm_mode = True

File: frontend/src/types/hr_agricole.ts
import { UUID } from './common';

// Types de base
export type TypePersonnel = 'PERMANENT' | 'SAISONNIER' | 'TEMPORAIRE' | 'STAGIAIRE';

export type SpecialiteAgricole = 'CULTURE' | 'ELEVAGE' | 'MARAICHAGE' | 'ARBORICULTURE' | 'MAINTENANCE' | 'LOGISTIQUE';

export type NiveauCompetence = 'DEBUTANT' | 'INTERMEDIAIRE' | 'AVANCE' | 'EXPERT';

export type TypeCertification = 'PHYTOSANITAIRE' | 'SECURITE' | 'CONDUITE_ENGINS' | 'BIO' | 'QUALITE';

// Interfaces principales
export interface CompetenceAgricole {
  id: UUID;
  employe_id: UUID;
  specialite: SpecialiteAgricole;
  niveau: NiveauCompetence;
  cultures: string[];
  equipements: string[];
  date_acquisition: string;
  date_mise_a_jour?: string;
  validite?: string;
  commentaire?: string;
  donnees_supplementaires: Record<string, any>;
}

export interface CertificationAgricole {
  id: UUID;
  competence_id: UUID;
  type_certification: TypeCertification;
  organisme: string;
  numero?: string;
  date_obtention: string;
  date_expiration?: string;
  niveau?: string;
  document_id?: UUID;
  commentaire?: string;
  donnees_supplementaires: Record<string, any>;
}

export interface AffectationParcelle {
  id: UUID;
  employe_id: UUID;
  parcelle_id: UUID;
  date_debut: string;
  date_fin?: string;
  role: string;
  responsabilites: string[];
  objectifs: Record<string, any>;
  restrictions_meteo: string[];
  equipements_requis: string[];
  commentaire?: string;
  donnees_supplementaires: Record<string, any>;
}

export interface ConditionTravailAgricole {
  id: UUID;
  employe_id: UUID;
  date: string;
  temperature?: number;
  humidite?: number;
  precipitation: boolean;
  vent?: number;
  exposition_soleil?: number;
  charge_physique?: number;
  equipements_protection: string[];
  incidents: Array<{
    type: string;
    duree?: number;
    mesures_prises: string[];
  }>;
  commentaire?: string;
  donnees_supplementaires: Record<string, any>;
}

export interface FormationAgricole {
  id: UUID;
  formation_id: UUID;
  specialite: SpecialiteAgricole;
  cultures_concernees: string[];
  equipements_concernes: string[];
  conditions_meteo: {
    temperature_min?: number;
    temperature_max?: number;
    conditions_optimales?: string[];
  };
  pratiques_specifiques: string[];
  evaluation_terrain: boolean;
  resultats_evaluation: {
    pratique?: number;
    theorie?: number;
    [key: string]: any;
  };
  commentaire?: string;
  donnees_supplementaires: Record<string, any>;
}

export interface EvaluationAgricole {
  id: UUID;
  evaluation_id: UUID;
  performances_cultures: Record<string, {
    rendement?: number;
    qualite?: number;
    techniques?: string[];
    [key: string]: any;
  }>;
  maitrise_equipements: Record<string, NiveauCompetence>;
  respect_securite: {
    epi?: number;
    procedures?: number;
    [key: string]: any;
  };
  adaptabilite_meteo: Record<string, NiveauCompetence>;
  gestion_ressources: Record<string, number>;
  qualite_travail: {
    precision?: number;
    rapidite?: number;
    [key: string]: any;
  };
  commentaire?: string;
  donnees_supplementaires: Record<string, any>;
}

// Types pour la création/mise à jour
export type CompetenceAgricoleCreate = Omit<CompetenceAgricole, 'id'>;
export type CompetenceAgricoleUpdate = Partial<CompetenceAgricoleCreate>;

export type CertificationAgricoleCreate = Omit<CertificationAgricole, 'id'>;
export type CertificationAgricoleUpdate = Partial<CertificationAgricoleCreate>;

export type AffectationParcelleCreate = Omit<AffectationParcelle, 'id'>;
export type AffectationParcelleUpdate = Partial<AffectationParcelleCreate>;

export type ConditionTravailAgricoleCreate = Omit<ConditionTravailAgricole, 'id'>;
export type ConditionTravailAgricoleUpdate = Partial<ConditionTravailAgricoleCreate>;

export type FormationAgricoleCreate = Omit<FormationAgricole, 'id'>;
export type FormationAgricoleUpdate = Partial<FormationAgricoleCreate>;

export type EvaluationAgricoleCreate = Omit<EvaluationAgricole, 'id'>;
export type EvaluationAgricoleUpdate = Partial<EvaluationAgricoleCreate>;

// Types utilitaires
export interface ValidationCompetence {
  valide: boolean;
}

export interface CompetenceARenouveler extends CompetenceAgricole {
  jours_restants: number;
}

export interface CertificationARenouveler extends CertificationAgricole {
  jours_restants: number;
}

// Types pour les statistiques
export interface StatistiquesCompetences {
  total: number;
  par_specialite: Record<SpecialiteAgricole, number>;
  par_niveau: Record<NiveauCompetence, number>;
  a_renouveler: number;
}

export interface StatistiquesCertifications {
  total: number;
  par_type: Record<TypeCertification, number>;
  valides: number;
  expirees: number;
  a_renouveler: number;
}

export interface StatistiquesFormations {
  total: number;
  par_specialite: Record<SpecialiteAgricole, number>;
  reussies: number;
  en_cours: number;
  planifiees: number;
}

export interface StatistiquesConditionsTravail {
  moyenne_temperature: number;
  moyenne_humidite: number;
  jours_precipitation: number;
  moyenne_charge_physique: number;
  incidents_total: number;
  incidents_par_type: Record<string, number>;
}

// Types pour les filtres
export interface FiltresRHAgricole {
  specialites?: SpecialiteAgricole[];
  niveaux?: NiveauCompetence[];
  types_certification?: TypeCertification[];
  date_debut?: string;
  date_fin?: string;
  valides_seulement?: boolean;
  a_renouveler?: boolean;
  delai_renouvellement?: number;
}

// Types pour le tri
export type ChampTriRHAgricole = 
  | 'date_acquisition'
  | 'date_obtention'
  | 'date_debut'
  | 'niveau'
  | 'specialite'
  | 'type_certification';

export interface OptionsTriRHAgricole {
  champ: ChampTriRHAgricole;
  ordre: 'asc' | 'desc';
}

// Types pour la pagination
export interface PaginationRHAgricole {
  page: number;
  limite: number;
  total: number;
}
